Feat/faucet attach 2

meowsbits opened the pull request:
Supersedes #259. 

- Creates `-attach` flag allowing the faucet program to _not_ run it's own client, instead attaching to an existing one via JSONRPC. Target client API needs to have `eth` and `net` API modules exposed.

- (https://github.com/etclabscore/core-geth/pull/269/commits/c99ed4582f30ed2ca289d43419ffd09a2639be40) _Disables_ `NoDiscovery` setting for managed client P2P configuration. Allowing discovery allows the faucet to not rely exclusively on manually configured bootnodes. Left commented because this differs from upstream... and I'm not sure at this point what the rationale was for disabling p2p discovery in the first place.

meowsbits committed:
main,ethclient,ethstats: init faucet attach feature

Rel https://github.com/etclabscore/core-geth/issues/258

This begins implementation of the ref'd feature proposal.

Currently 'ethstats' support is removed because
that package relies heavily on access to a *node.Node.
Eventually I want to move that packages demands to
something that can be satisfied by an API instead.

Date: 2020-12-21 12:16:33-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Merge branch 'master' into feat/faucet-attach

Conflicts:
	cmd/faucet/faucet.go

meowsbits committed:
ethstats: revert to use consensus.Engine interface

I'm reverting the HeaderAuthorGetter interface
solely to reduce diff size and complexity, both
for this PR and in general.

It doesn't mean I like it.

Date: 2020-12-28 07:47:24-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: handle TODOs around faucet attachery

Work was pending resolution of ethstats reporting
in conjunction with node attaching.

This has been resolved conceptually by
make them exclusive; you cannot activate ethstats
for the faucet's target attach node (since that
node can handle ethstats reporting all by itself).

Date: 2020-12-28 08:05:24-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: fix typo in conditional decl

Date: 2020-12-28 08:08:25-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: create flag use audit function

Make sure flags aren't used in impossible ways.

Date: 2020-12-28 08:17:13-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: add flag exclusive case to audit (-genesis, -attach)

Date: 2020-12-28 08:24:18-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: add info log lines around genesis/attach features

Date: 2020-12-28 08:25:09-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
ethstats: revert unrelated changes

Date: 2020-12-28 08:37:41-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits commented on cmd/faucet/faucet.go:
```diff
@@ -286,7 +306,26 @@ func main() {
+	// See the difference in implementation between ethapi/api.go and eth/api.go #ChainID() methods.
+	// There's an issue open about this somewhere at ethereum/xxx.
+	// This could be resolved by creating a -chainid flag to use as a fallback.
+	chainID := uint64(0)
```
:eyes: (see comment)

ziogaschr replied:
`-chainid` sounds good to have as a fallback.

Maybe we can request the genesis hash from the attached node and compare it locally? (till EIP155)

meowsbits replied:
Yeap, that'll work. I just wrote the code for it (see https://github.com/etclabscore/core-geth/pull/269/commits/b78074862257c7943e18df3912893a01d7ba4a6e and +1), but am now getting the feeling of fix-it-with-a-flag/feature-creep. The only thing this is used for is configuring the data directory namespace. #268 includes a change adding a `-datadir` flag that would provide another way around this potential ambiguity for a user. Otherwise, this ambiguity would only be an issue for users attempting to attach the `faucet` to a (currently) very-unsynced client. I'm starting to feel inclined to just `if chainID == 0 && { log.Crit... }`.

ziogaschr replied:
I like the way https://github.com/etclabscore/core-geth/commit/b78074862257c7943e18df3912893a01d7ba4a6e + 1 handles it. Bravo!!!

meowsbits committed:
ethstats: (lint) goimports

Date: 2020-12-28 10:10:51-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: disable NoDiscovery (enable discovery)

Faucet behavior previously effectively demanded
user configuration of bootnodes manually
to enable peering.

This disables NoDiscovery, thus
enabling discovery.

Date: 2020-12-28 10:19:05-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: (lint) goimports

Date: 2020-12-28 10:22:20-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: configure datadir with genesis hash too

As noted in the comment on line 310, when
attaching to a remote client it may be that the
API says that chain id is 0 when the client
is not yet synced past the EIP155 activation
block height.

Requesting and switching on the genesis block
hash fixes this issue except for the foundation/classic
corner case (same genesis, different chain ids).

Date: 2020-12-29 05:14:04-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: add -attach.chainid to provide accessible fallback value

This value would be useful and necessary only if
the target client were unsynced past the EIP155
activation height.

ChainID is used to disambiguate ETH/ETC chains.

Date: 2020-12-29 05:30:04-06:00
Signed-off-by: meows <b5c6@protonmail.com>

iquidus commented on cmd/faucet/faucet.go:
```diff
@@ -629,7 +725,15 @@ func (f *faucet) apiHandler(w http.ResponseWriter, r *http.Request) {
 			tx := types.NewTransaction(f.nonce+uint64(len(f.reqs)), address, amount, 21000, f.price, nil)
-			signed, err := f.keystore.SignTx(f.account, tx, f.config.GetChainID())
+
+			// FIXME(meowsbits): Getting the chain id more than once is redundant and can be optimized.
```
is there a solution for this in mind?

meowsbits replied:
Probably there are at least a few ways of handling it. One that comes first to mind is to add a field on `faucet`, eg. `faucet.chainID` since we know that once it's set to a nonzero value it won't change, and the `faucet` is only designed to handle a single network. But IMO this is a nice-to-have optimization since the request will only be made for each faucet transaction... which isn't more than can be expected for a "standard" transaction-making program (at least it's not getting called at every block or anything like that). Because of this I'm OK pushing to a later optimization if need/desire calls for it.

ziogaschr commented on cmd/faucet/faucet.go:
```diff
@@ -176,107 +273,33 @@ func main() {
-		} else {
-			log.Error("Failed to parse bootnode URL", "url", boot, "err", err)
+	} else {
+		log.Info("Attaching faucet to running client")
```
@meowsbits this message is a bit misleading. While testing #268 without using neither `-attach` nor a chain `-chain.*` flag. It get's in here saying it failed to attach. While the truth is that it failed to peek a chain.

meowsbits replied:
Nice catch. Please see https://github.com/etclabscore/core-geth/pull/275.
